server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+
+  protected String getProviderName() {
+    return getConfigurationValue(attributePrefix() + "." + providerNameConfigName);
+  }
+
+  protected String getClientID() {
+    return getConfigurationValue(attributePrefix() + "." + clientIDConfigName);
+  }
+
+  protected String getClientSecret() {
+    return getConfigurationValue(attributePrefix() + "." + clientSecretConfigName);
+  }
+
+  protected String getDiscoveryURI() {
+    return getConfigurationValue(attributePrefix() + "." + discoveryURIConfigName);
+  }
+
+  protected String getResponseMode() {
+    return getConfigurationValue(attributePrefix() + "." + responseModeConfigName);
+  }
+
+  protected String getResponseType() {
+    // Our local fake IDCS doesn't support 'token' auth.
+    if (baseUrl.contains("localhost:")) {
+      return "id_token";
+    }
+    return "id_token token";
+  }
+
+  protected String[] getExtraScopes() {
+    return getConfigurationValue(attributePrefix() + "." + extraScopesConfigName).split(" ");
+  }
+
+  protected String getCallbackURL() {
+    return baseUrl + "/callback";
+  }
+
+  protected String getScopes() {
+    // Scopes are the other things that we want from the OIDC endpoint
+    // (needs to also be configured on provider side).
+    String[] extraScopes = getExtraScopes();
+
+    ArrayList<String> allClaims = new ArrayList<>();
+    Collections.addAll(allClaims, defaultScopes);
+    Collections.addAll(allClaims, extraScopes);
+    return String.join(" ", allClaims);
+  }
+
+  @Override
+  public OidcClient get() {
+    String clientID = Objects.toString(getClientID(), "");
+    String clientSecret = Objects.toString(getClientSecret(), "");
+    String discoveryURI = Objects.toString(getDiscoveryURI(), "");
+    String responseMode = Objects.toString(getResponseMode(), "");
+    String responseType = Objects.toString(getResponseType(), "");
+    String callbackURL = Objects.toString(getCallbackURL(), "");
+    String providerName = Objects.toString(getProviderName(), ""); // optional
+    String scopes = Objects.toString(getScopes(), "");
+    if (clientID.isBlank()
+        || clientSecret.isBlank()
+        || discoveryURI.isBlank()
+        || responseMode.isBlank()
+        || responseType.isBlank()
+        || callbackURL.isBlank()
+        || baseUrl.isBlank()) {
+      logger.error(
+          String.format(
+              "Can't get OIDC client - Missing Provider data:\n"
+                  + " baseUrl=%s\n"
+                  + " clientID=%s \n"
+                  + " clientSecret=%s \n"
+                  + " discoveryURI=%s \n"
+                  + " responseMode=%s \n"
+                  + " responseType=%s \n"
+                  + " callbackURL=%s \n"
+                  + " providerName=%s",
+              baseUrl,
+              clientID,
+              clientSecret,
+              discoveryURI,
+              responseMode,
+              responseType,
+              callbackURL,
+              providerName));
+      return null;
```

Comment:
Not blocking; Consider returning an Optional<OidcClient> rather than a null value (for internal codebase consistency).

Reply:
blocking: either return Optional or throw, generally avoid returning `null` whenever possible